test/type-errors/Deriving.hs

```diff
@@ -0,0 +1,404 @@
+ GHC.Generic.
+-}
+newtype NewtypeWrappedR_NiceNamedM a
+  = NewtypeWrappedR_NiceNamedM a
+  deriving newtype (GHC.Generic)
+
+instance IsRoute (NewtypeWrappedR_NiceNamedM a) where
+  type RouteModel (NewtypeWrappedR_NiceNamedM a) = NiceNamedM () ()
+  routePrism = undefined
+  routeUniverse = undefined
+
+-- * Test cases
+
+#define ENABLE_SPEC
+
+----------------------------------------
+-- Subroute verification
+----------------------------------------
+
+#undef ENABLE_SPEC
+#ifdef ENABLE_SPEC
+routeSpec "subroutes should not have constructors with multiple fields"
+  (badRoute ''() ''())
+  [t|
+    '[ WithModel (NiceNamedM () ())
+     ]
+  |]
+  [r|
+MultiRoute: too many arguments
+  |]
+#endif
+#define ENABLE_SPEC
+
+-----------------------------------------
+
+#undef ENABLE_SPEC
+#ifdef ENABLE_SPEC
+routeSpec "'WithSubRoutes' list should not be shorter than number of route constructors"
+  (niceRoute ''() ''())
+  [t|
+    '[ WithModel (NiceNamedM () ())
+     , WithSubRoutes '[ () ]
+     ]
+  |]
+  [r|
+'WithSubRoutes' is missing subroutes for:
+
+  '[ '[()]]
+  |]
+#endif
+#define ENABLE_SPEC
+
+-----------------------------------------
+
+#undef ENABLE_SPEC
+#ifdef ENABLE_SPEC
+routeSpec "'WithSubRoutes' list should not be longer than number of route constructors"
+  (niceRoute ''() ''())
+  [t|
+    '[ WithModel (NiceNamedM () ())
+     , WithSubRoutes '[ (), (), () ]
+     ]
+  |]
+  [r|
+'WithSubRoutes' has extra unnecessary types:
+
+  '[()]
+  |]
+#endif
+#define ENABLE_SPEC
+
+------------------------------------------
+
+#undef ENABLE_SPEC
+#ifdef ENABLE_SPEC
+routeSpec "constructors should either be empty or contain () when 'WithSubRoutes' specifies ()"
+  (niceRoute ''Int ''())
+  [t|
+    '[ WithModel (NiceNamedM () ())
+     , WithSubRoutes '[ (), () ]
+     ]
+  |]
+  [r|
+A 'WithSubRoutes' entry is '()' instead of the expected:
+'[Int]
+  |]
+#endif
+#define ENABLE_SPEC
+
+-------------------------------------------
+
+#undef ENABLE_SPEC
+#ifdef ENABLE_SPEC
+routeSpec "subroute types that are nonisomorphic to what is specified in 'WithSubRoutes' should be illegal"
+  (niceRoute ''() ''())
+  [t|
+    '[ WithModel (NiceNamedM () ())
+     , WithSubRoutes '[ (), Bool ]
+     ]
+  |]
+  [r|
+A 'WithSubRoutes' type:
+
+  Bool
+
+is not isomorphic to the corresponding route constructor type:
+
+  '[()]
+
+  |]
+#endif
+#define ENABLE_SPEC
+
+-------------------------------------------
+
+routeSpec
+  "subroute types that are the same as what is specified in 'WithSubRoutes' should typecheck"
+  (niceRoute ''() ''())
+  [t|
+    '[ WithModel (NiceNamedM () ())
+     , WithSubRoutes '[(), ()]
+     ]
+    |]
+  [r|
+  |]
+
+-------------------------------------------
+
+#undef ENABLE_SPEC
+#ifdef ENABLE_SPEC
+routeSpec "subroute types that are an unwrapped representation of what is specified in 'WithSubRoutes' should typecheck | ( empty constructor <-> () ) special case"
```

Comment:
This spec says it "should typecheck", yet when you uncomment the undef above a type error is thrown:

<img width="521" alt="image" src="https://user-images.githubusercontent.com/3998/181582759-8c77d04f-2dd5-4e3c-bafc-714852661879.png">

I must admit that I'm feeling somewhat uncomfortable with this whole `niceRoute` abstraction from a long-term maintenance perspective. It is not immediately clear what route type is being used and what deriving it does to it. In contrast, if each `routeSpec` defined its own (unique) route and model, it would become immediately obvious.

Reply:
Yep the empty <-> () unwrap case was a regression from #120 , next PR will fix this.

> I must admit that I'm feeling somewhat uncomfortable with this whole niceRoute abstraction from a long-term maintenance perspective. It is not immediately clear what route type is being used and what deriving it does to it. In contrast, if each routeSpec defined its own (unique) route and model, it would become immediately obvious.

Possible workarounds off the top of my head:
a) I could dump the splices from this module and paste them in; some route names will be very long though.
b) Document output of niceRoute/badRoute better and ask people to dump splices if they need to for debugging purposes.

Reply:
Oops, sorry about this. Been swamped with other stuff (2 pending PRs on emanote as well). I'll get to it this week.

But meanwhile, if you could show what TH actually generates for say the above case it would be helpful. Once I have the time, I was planning on doing this myself for a few cases to assess to what extent TH actually helps us here.

Reply:
@RiugaBachi I removed all the TH-based abstraction, and the resulting code is not as duplicate-ridden as we assumed.

Are you familiar with the [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))? For future work, I'd highly advise avoiding any "premature refactoring" (to coin a term), and instead, give yourself permission to write ... what seems like duplicate code. Often a better abstraction comes along only after you've seen N duplicates (the rule of three says n=3); sometimes (as with this case), you may find that an abstraction is not even needed.

Furthermore, in this particular case, I find the tests to be far more readable because all the types/code exist locally within the `#ifdef ... #endif` block.